kaldb/src/main/java/com/slack/kaldb/preprocessor/PreprocessorService.java

```diff
@@ -0,0 +1,285 @@
+    LOG.info("Stopping preprocessor service");
+    if (kafkaStreams != null) {
+      kafkaStreams.close(DEFAULT_START_STOP_DURATION);
+    }
+    if (kafkaStreamsMetrics != null) {
+      kafkaStreamsMetrics.close();
+    }
+    serviceMetadataStore.removeListener(this::load);
+    LOG.info("Preprocessor service closed");
+  }
+
+  /**
+   * Configures and starts a KafkaStream processor, based off of the cached ServiceMetadataStore.
+   * This method is reentrant, and will restart any existing KafkaStream processors. Access to this
+   * must be synchronized if using this method as part of a listener.
+   */
+  public synchronized void load() {
+    configReloadCounter.increment();
+    LOG.info("Loading new Kafka stream processor config");
+    if (kafkaStreams != null) {
+      LOG.info("Closing existing Kafka stream processor");
+      kafkaStreams.close();
+      kafkaStreams.cleanUp();
+    }
+    if (kafkaStreamsMetrics != null) {
+      kafkaStreamsMetrics.close();
+    }
+
+    // only attempt to register stream processing on valid service configurations
+    List<ServiceMetadata> serviceMetadataToProcess =
+        filterValidServiceMetadata(serviceMetadataStore.listSync());
+
+    if (serviceMetadataToProcess.size() > 0) {
+      Topology topology =
+          buildTopology(
+              serviceMetadataToProcess,
+              rateLimiter,
+              upstreamTopics,
+              downstreamTopic,
+              dataTransformer);
+      kafkaStreams = new KafkaStreams(topology, kafkaProperties);
+      kafkaStreamsMetrics = new KafkaStreamsMetrics(kafkaStreams);
+      kafkaStreamsMetrics.bindTo(meterRegistry);
+      kafkaStreams.start();
+      LOG.info("Kafka stream processor config loaded successfully");
+    } else {
+      LOG.info(
+          "No valid service configurations found to process - will retry on next service configuration update");
+    }
+  }
+
+  /**
+   * Builds a KafkaStream Topology with multiple source topics, targeting a single sink. Applies
+   * rate limits per-service if required, and uses the service configured target topic sink and data
+   * transformer.
+   */
+  protected static Topology buildTopology(
+      List<ServiceMetadata> serviceMetadataList,
+      PreprocessorRateLimiter rateLimiter,
+      List<String> upstreamTopics,
+      String downstreamTopic,
+      String dataTransformer) {
+    Preconditions.checkArgument(
+        !serviceMetadataList.isEmpty(), "service metadata list must not be empty");
+    Preconditions.checkArgument(upstreamTopics.size() > 0, "upstream topic list must not be empty");
+    Preconditions.checkArgument(!downstreamTopic.isEmpty(), "downstream topic must not be empty");
+    Preconditions.checkArgument(!dataTransformer.isEmpty(), "data transformer must not be empty");
+
+    StreamsBuilder builder = new StreamsBuilder();
+
+    ValueMapper<byte[], Iterable<Trace.Span>> valueMapper =
+        PreprocessorValueMapper.byteArrayToTraceSpans(dataTransformer);
```

Comment:
nit: add a new line for better readability

Reply:
Fixed.